Change Substs to type alias for Slice<Kind> for interning

dikaiosune opened the pull request:
This changes the definition of `librustc::ty::subst::Substs` to be a type alias to `Slice<Kind>`. `Substs` was already interned, but can now make use of the efficient `PartialEq` and `Hash` impls on `librustc::ty::Slice`.

I'm working on collecting some timing data for this, will update when it's done.

I chose to leave the impls on `Substs<'tcx>` even though it's now just a type alias to `Slice<Kind<'tcx>>` because it has the smallest footprint on other portions of the compiler which depend on its API. It turns out to be a pretty huge diff if you change where Substs's methods live :smile:. That said, I'm not necessarily sure it's the *best* implementation but it's probably the easiest/smallest to review.

Many thanks to @eddyb for both suggesting this as a project for learning more about the compiler, and the tireless ~~handholding~~ mentorship he provided.

rust-highfive commented:
Thanks for the pull request, and welcome! The Rust team is excited to review your changes, and you should hear from @nrc (or someone else) soon.

If any changes to this PR are deemed necessary, please add them as extra commits. This ensures that the reviewer can see what has changed since they last reviewed the code. Due to the way GitHub handles out-of-date commits, this should also make it reasonably obvious what issues have or haven't been addressed. Large or tricky changes may require several passes of review and changes.

Please see [the contribution instructions](https://github.com/rust-lang/rust/blob/master/CONTRIBUTING.md) for more information.

eddyb commented on src/librustc/ty/subst.rs:
```diff
@@ -206,82 +199,81 @@ impl<'a, 'gcx, 'tcx> Substs<'tcx> {
     pub fn params(&self) -> &[Kind<'tcx>] {
-        &self.params
+        &self
     }
```
You can probably remove this since it's a noop.

dikaiosune replied:
I definitely can, but that will increase the size of the diff since many callers currently rely on the params method. Worth removing?

eddyb replied:
Can do it later, I suppose. At least put a `// FIXME` comment on something so people don't get confused - oh and it can be just `self`.

dikaiosune replied:
Done

eddyb commented on src/librustc/ty/subst.rs:
```diff
@@ -305,27 +297,22 @@ impl<'a, 'gcx, 'tcx> Substs<'tcx> {
-        folder.tcx().mk_substs(Substs {
-            params: params
-        })
+        let params = self.iter().map(|k| k.fold_with(folder)).collect();
```
Unnecessary `.collect()`.

dikaiosune replied:
Perhaps you can help me think of a better solution -- `Substs::new` takes `IntoIterator<Item=Kind>`, which is what we'd really want here. But the closure in the `map` call here borrows `folder` mutably, and `Substs::new` wants a `TyCtxt` from the folder, trying to borrow that immutably. And trying to create a simultaneous pair of mutable/immutable borrows obviously creates a borrow error.

The "workaround" I decided on was to allocate here (it has to happen somewhere before interning), and then call the intern function directly instead of going through the iteration-based approach. This should still only allocate the `Vec<Kind>` the once, but its an ugly bypass of the "right" interface.

eddyb replied:
The `tcx` doesn't borrow the folder, it's copied out (I hope).

eddyb commented on src/librustc_trans/trans_item.rs:
```diff
@@ -359,7 +338,7 @@ impl<'a, 'tcx> TransItem<'tcx> {
                 format!("Fn({:?}, {})",
                          instance.def,
-                         instance.substs as *const _ as usize)
+                         instance.substs.as_ptr() as usize)
```
Need to also put the length here, unless this is not used for anything sensitive - cc @michaelwoerister.

dikaiosune replied:
Can't hurt, right?

dikaiosune replied:
Done (hopefully correctly?)

eddyb replied:
I take it back, the length is determined by `instance.def` so this chance is not necessary.

dikaiosune replied:
OK, reverted.

dikaiosune commented:
I think I've addressed the nits here. On my machine, I saw a roughly 4.5% improvement to the overall time of non-LLVM passes when building rustc-benchmarks (when aggregated by pass as a geometric mean). That said, digging into individual passes on individual crates shows a pretty high degree of variability on my machine between builds, and while it seems like the overall improvement is outside of the error bars there, I'm not a statistics expert :).

I suspect that many Substs may be less than 2 Kinds long, in which case @arielb1's suggestion from IRC of doing something like SSO may be good to investigate.

Are there any other timing measurements that I should collect?

eddyb commented:
The SSO sugestion is a subset of what I want to do (small stack array plus slice arena alocations) and you canct just reuse an existing abstraction, they're a bit hard to optimize at all.

eddyb commented:
@bors r+

bors commented:
:pushpin: Commit 48b3dd1 has been approved by `eddyb`

<!-- @bors r=eddyb 48b3dd11f59f48819031206ee2b3ab98ceae1550 -->

bors commented:
:hourglass: Testing commit 48b3dd1 with merge 7aa6e06...

bors commented:
:broken_heart: Test failed - [auto-linux-cross-opt](https://buildbot.rust-lang.org/builders/auto-linux-cross-opt/builds/3908)

arielb1 commented:
@bors retry

alexcrichton commented:
@bors: retry

bors commented:
:hourglass: Testing commit 48b3dd1 with merge 8e05e7e...

bors commented:
:sunny: Test successful - [auto-linux-32-nopt-t](https://buildbot.rust-lang.org/builders/auto-linux-32-nopt-t/builds/10767), [auto-linux-32-opt](https://buildbot.rust-lang.org/builders/auto-linux-32-opt/builds/10823), [auto-linux-64-cargotest](https://buildbot.rust-lang.org/builders/auto-linux-64-cargotest/builds/2054), [auto-linux-64-cross-freebsd](https://buildbot.rust-lang.org/builders/auto-linux-64-cross-freebsd/builds/1879), [auto-linux-64-debug-opt](https://buildbot.rust-lang.org/builders/auto-linux-64-debug-opt/builds/4039), [auto-linux-64-nopt-t](https://buildbot.rust-lang.org/builders/auto-linux-64-nopt-t/builds/10766), [auto-linux-64-opt](https://buildbot.rust-lang.org/builders/auto-linux-64-opt/builds/10720), [auto-linux-64-opt-rustbuild](https://buildbot.rust-lang.org/builders/auto-linux-64-opt-rustbuild/builds/2696), [auto-linux-64-x-android-t](https://buildbot.rust-lang.org/builders/auto-linux-64-x-android-t/builds/10694), [auto-linux-cross-opt](https://buildbot.rust-lang.org/builders/auto-linux-cross-opt/builds/3923), [auto-linux-musl-64-opt](https://buildbot.rust-lang.org/builders/auto-linux-musl-64-opt/builds/5871), [auto-mac-32-opt](https://buildbot.rust-lang.org/builders/auto-mac-32-opt/builds/10804), [auto-mac-64-opt](https://buildbot.rust-lang.org/builders/auto-mac-64-opt/builds/10779), [auto-mac-64-opt-rustbuild](https://buildbot.rust-lang.org/builders/auto-mac-64-opt-rustbuild/builds/2724), [auto-mac-cross-ios-opt](https://buildbot.rust-lang.org/builders/auto-mac-cross-ios-opt/builds/1903), [auto-win-gnu-32-opt](https://buildbot.rust-lang.org/builders/auto-win-gnu-32-opt/builds/5903), [auto-win-gnu-32-opt-rustbuild](https://buildbot.rust-lang.org/builders/auto-win-gnu-32-opt-rustbuild/builds/2759), [auto-win-gnu-64-opt](https://buildbot.rust-lang.org/builders/auto-win-gnu-64-opt/builds/5868), [auto-win-msvc-32-opt](https://buildbot.rust-lang.org/builders/auto-win-msvc-32-opt/builds/5063), [auto-win-msvc-64-cargotest](https://buildbot.rust-lang.org/builders/auto-win-msvc-64-cargotest/builds/2074), [auto-win-msvc-64-opt](https://buildbot.rust-lang.org/builders/auto-win-msvc-64-opt/builds/5882), [auto-win-msvc-64-opt-rustbuild](https://buildbot.rust-lang.org/builders/auto-win-msvc-64-opt-rustbuild/builds/2752)
Approved by: eddyb
Pushing 8e05e7ee3c19a5594b79d67c8390cef78970be7c to master...
